File: packages/SystemUI/src/com/android/systemui/statusbar/notification/stack/NotificationSectionsManager.java
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.systemui.statusbar.notification.stack;

import static com.android.systemui.statusbar.notification.stack.NotificationStackScrollLayout.ROWS_GENTLE;

import android.annotation.Nullable;
import android.content.Intent;
import android.provider.Settings;
import android.view.LayoutInflater;
import android.view.View;

import com.android.internal.annotations.VisibleForTesting;
import com.android.systemui.R;
import com.android.systemui.plugins.ActivityStarter;
import com.android.systemui.plugins.statusbar.StatusBarStateController;
import com.android.systemui.statusbar.StatusBarState;
import com.android.systemui.statusbar.notification.row.ActivatableNotificationView;
import com.android.systemui.statusbar.notification.row.ExpandableNotificationRow;
import com.android.systemui.statusbar.policy.ConfigurationController;
import com.android.systemui.statusbar.policy.ConfigurationController.ConfigurationListener;

/**
 * Manages the boundaries of the two notification sections (high priority and low priority). Also
 * shows/hides the headers for those sections where appropriate.
 *
 * TODO: Move remaining sections logic from NSSL into this class.
 */
class NotificationSectionsManager implements StackScrollAlgorithm.SectionProvider {
    private final NotificationStackScrollLayout mParent;
    private final ActivityStarter mActivityStarter;
    private final StatusBarStateController mStatusBarStateController;
    private final ConfigurationController mConfigurationController;
    private final boolean mUseMultipleSections;

    private boolean mInitialized = false;
    private SectionHeaderView mGentleHeader;
    private SectionHeaderView mAlertHeader;
    private boolean mGentleHeaderVisible = false;
    private boolean mAlertHeaderVisible = false;
    @Nullable private ExpandableNotificationRow mFirstGentleNotif;
    @Nullable private ExpandableNotificationRow mFirstAlertNotif;

    NotificationSectionsManager(
            NotificationStackScrollLayout parent,
            ActivityStarter activityStarter,
            StatusBarStateController statusBarStateController,
            ConfigurationController configurationController,
            boolean useMultipleSections) {
        mParent = parent;
        mActivityStarter = activityStarter;
        mStatusBarStateController = statusBarStateController;
        mConfigurationController = configurationController;
        mUseMultipleSections = useMultipleSections;
    }

    /** Must be called before use. */
    void initialize(LayoutInflater layoutInflater) {
        if (mInitialized) {
            throw new IllegalStateException("NotificationSectionsManager already initialized");
        }
        mInitialized = true;
        reinflateViews(layoutInflater);
        mConfigurationController.addCallback(mConfigurationListener);
    }

    void reinflateViews(LayoutInflater layoutInflater) {
        mGentleHeader = reinflateView(layoutInflater, mGentleHeader,
                R.string.notification_section_header_gentle);
        mAlertHeader = reinflateView(layoutInflater, mAlertHeader,
                R.string.notification_section_header_alerting);
    }

    /**
     * Reinflates the entire notification header, including all decoration views.
     */
    SectionHeaderView reinflateView(LayoutInflater layoutInflater,
            SectionHeaderView sectionHeader, int labelId) {
        int oldPos = -1;
        if (sectionHeader != null) {
            if (sectionHeader.getTransientContainer() != null) {
                sectionHeader.getTransientContainer().removeView(sectionHeader);
            } else if (sectionHeader.getParent() != null) {
                oldPos = mParent.indexOfChild(sectionHeader);
                mParent.removeView(sectionHeader);
            }
        }

        sectionHeader = (SectionHeaderView) layoutInflater.inflate(
                R.layout.status_bar_notification_section_header, mParent, false);
        sectionHeader.setLabelText(mParent.getContext().getResources().getString(labelId));

        if (oldPos != -1) {
            mParent.addView(sectionHeader, oldPos);
        }

        return sectionHeader;
    }

    @Override
    public boolean beginsSection(View view) {
        return view == getFirstLowPriorityChild() || view == getFirstHighPriorityChild();
    }

    /**
     * Should be called whenever notifs are added, removed, or updated. Updates section boundary
     * bookkeeping and adds/moves/removes section headers if appropriate.
     */
    void updateSectionBoundaries() {
        if (!mUseMultipleSections) {
            return;
        }

        mGentleHeaderVisible = adjustHeaderVisibilityAndPosition(
                getFirstNotificationCategoryIndex(true), mGentleHeader, mGentleHeaderVisible);
        mAlertHeaderVisible = adjustHeaderVisibilityAndPosition(
                getFirstNotificationCategoryIndex(false), mAlertHeader, mAlertHeaderVisible);
    }

    private int getFirstNotificationCategoryIndex(boolean gentle) {
        final int n = mParent.getChildCount();
        for (int i = 0; i < n; i++) {
            View child = mParent.getChildAt(i);
            if (child instanceof ExpandableNotificationRow
                    && child.getVisibility() != View.GONE) {
                ExpandableNotificationRow row = (ExpandableNotificationRow) child;
                if (!gentle && row.getEntry().isTopBucket()) {
                    mFirstAlertNotif = row;
                    return i;
                } else if (gentle && !row.getEntry().isTopBucket()) {
                    mFirstGentleNotif = row;
                    return i;
                }
            }
        }

        return -1;
    }

    private boolean adjustHeaderVisibilityAndPosition(int firstNotifIndex,
            SectionHeaderView sectionHeader, boolean headerCurrentlyVisible) {
        final boolean showHeader = firstNotifIndex != -1
                && mStatusBarStateController.getState() != StatusBarState.KEYGUARD;
        final int currentHeaderIndex = mParent.indexOfChild(sectionHeader);

        if (!showHeader) {
            if (headerCurrentlyVisible) {
                headerCurrentlyVisible = false;
                mParent.removeView(sectionHeader);
                sectionHeader.setVisible(false, false);
            }
        } else {
            if (!headerCurrentlyVisible) {
                headerCurrentlyVisible = true;
                // If the header is animating away, it will still have a parent, so detach it first
                // TODO: We should really cancel the active animations here. This will happen
                // automatically when the view's intro animation starts, but it's a fragile link.
                if (sectionHeader.getTransientContainer() != null) {
                    sectionHeader.getTransientContainer().removeTransientView(sectionHeader);
                    sectionHeader.setTransientContainer(null);
                }
                mParent.addView(sectionHeader, firstNotifIndex);
                sectionHeader.setVisible(true, false);
            } else if (currentHeaderIndex != firstNotifIndex - 1) {
                // Relocate the header to be immediately before the first child in the section
                int targetIndex = firstNotifIndex;
                if (currentHeaderIndex < firstNotifIndex) {
                    // Adjust the target index to account for the header itself being temporarily
                    // removed during the position change.
                    targetIndex--;
                }

                mParent.changeViewPosition(sectionHeader, targetIndex);
            }
        }

        return headerCurrentlyVisible;
    }

    /**
     * Updates the boundaries (as tracked by their first and last views) of the high and low
     * priority sections.
     *
     * @return {@code true} If the last view in the top section changed (so we need to animate).
     */
    boolean updateFirstAndLastViewsInSections(
            final NotificationSection highPrioritySection,
            final NotificationSection lowPrioritySection,
            ActivatableNotificationView firstChild,
            ActivatableNotificationView lastChild) {
        if (mUseMultipleSections) {
            ActivatableNotificationView previousLastHighPriorityChild =
                    highPrioritySection.getLastVisibleChild();
            ActivatableNotificationView previousFirstLowPriorityChild =
                    lowPrioritySection.getFirstVisibleChild();
            ActivatableNotificationView lastHighPriorityChild = getLastHighPriorityChild();
            ActivatableNotificationView firstLowPriorityChild = getFirstLowPriorityChild();
            if (lastHighPriorityChild != null && firstLowPriorityChild != null) {
                highPrioritySection.setFirstVisibleChild(firstChild);
                highPrioritySection.setLastVisibleChild(lastHighPriorityChild);
                lowPrioritySection.setFirstVisibleChild(firstLowPriorityChild);
                lowPrioritySection.setLastVisibleChild(lastChild);
            } else if (lastHighPriorityChild != null) {
                highPrioritySection.setFirstVisibleChild(firstChild);
                highPrioritySection.setLastVisibleChild(lastChild);
                lowPrioritySection.setFirstVisibleChild(null);
                lowPrioritySection.setLastVisibleChild(null);
            } else {
                highPrioritySection.setFirstVisibleChild(null);
                highPrioritySection.setLastVisibleChild(null);
                lowPrioritySection.setFirstVisibleChild(firstChild);
                lowPrioritySection.setLastVisibleChild(lastChild);
            }
            return lastHighPriorityChild != previousLastHighPriorityChild
                    || firstLowPriorityChild != previousFirstLowPriorityChild;
        } else {
            highPrioritySection.setFirstVisibleChild(firstChild);
            highPrioritySection.setLastVisibleChild(lastChild);
            return false;
        }
    }

    @VisibleForTesting
    SectionHeaderView getGentleHeaderView() {
        return mGentleHeader;
    }

    @Nullable
    private ActivatableNotificationView getFirstLowPriorityChild() {
        return mFirstGentleNotif;
    }

    private ActivatableNotificationView getFirstHighPriorityChild() {
        return mFirstAlertNotif;
    }

    @Nullable
    private ActivatableNotificationView getLastHighPriorityChild() {
        ActivatableNotificationView lastChildBeforeGap = null;
        int childCount = mParent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            View child = mParent.getChildAt(i);
            if (child.getVisibility() != View.GONE && child instanceof ExpandableNotificationRow) {
                ExpandableNotificationRow row = (ExpandableNotificationRow) child;
                if (!row.getEntry().isTopBucket()) {
                    break;
                } else {
                    lastChildBeforeGap = row;
                }
            }
        }
        return lastChildBeforeGap;
    }

    private final ConfigurationListener mConfigurationListener = new ConfigurationListener() {
        @Override
        public void onLocaleListChanged() {
            mGentleHeader.reinflateContents();
            mAlertHeader.reinflateContents();
        }
    };
}

File: packages/SystemUI/src/com/android/systemui/statusbar/notification/stack/NotificationSectionsManager.kt
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.systemui.statusbar.notification.stack

import android.annotation.ColorInt
import android.util.Log
import android.view.View
import com.android.internal.annotations.VisibleForTesting
import com.android.systemui.media.controls.ui.controller.KeyguardMediaController
import com.android.systemui.statusbar.notification.NotificationSectionsFeatureManager
import com.android.systemui.statusbar.notification.SourceType
import com.android.systemui.statusbar.notification.collection.render.MediaContainerController
import com.android.systemui.statusbar.notification.collection.render.SectionHeaderController
import com.android.systemui.statusbar.notification.dagger.AlertingHeader
import com.android.systemui.statusbar.notification.dagger.IncomingHeader
import com.android.systemui.statusbar.notification.dagger.PeopleHeader
import com.android.systemui.statusbar.notification.dagger.SilentHeader
import com.android.systemui.statusbar.notification.row.ExpandableNotificationRow
import com.android.systemui.statusbar.notification.row.ExpandableView
import com.android.systemui.statusbar.notification.stack.StackScrollAlgorithm.SectionProvider
import com.android.systemui.statusbar.policy.ConfigurationController
import com.android.systemui.util.foldToSparseArray
import javax.inject.Inject

/**
 * Manages section headers in the NSSL.
 *
 * TODO: Move remaining sections logic from NSSL into this class.
 */
class NotificationSectionsManager
@Inject
internal constructor(
    private val configurationController: ConfigurationController,
    private val keyguardMediaController: KeyguardMediaController,
    private val sectionsFeatureManager: NotificationSectionsFeatureManager,
    private val mediaContainerController: MediaContainerController,
    private val notificationRoundnessManager: NotificationRoundnessManager,
    @IncomingHeader private val incomingHeaderController: SectionHeaderController,
    @PeopleHeader private val peopleHeaderController: SectionHeaderController,
    @AlertingHeader private val alertingHeaderController: SectionHeaderController,
    @SilentHeader private val silentHeaderController: SectionHeaderController
) : SectionProvider {

    private val configurationListener =
        object : ConfigurationController.ConfigurationListener {
            override fun onLocaleListChanged() {
                reinflateViews()
            }
        }

    private lateinit var parent: NotificationStackScrollLayout
    private var initialized = false

    @VisibleForTesting
    val silentHeaderView: SectionHeaderView?
        get() = silentHeaderController.headerView

    @VisibleForTesting
    val alertingHeaderView: SectionHeaderView?
        get() = alertingHeaderController.headerView

    @VisibleForTesting
    val incomingHeaderView: SectionHeaderView?
        get() = incomingHeaderController.headerView

    @VisibleForTesting
    val peopleHeaderView: SectionHeaderView?
        get() = peopleHeaderController.headerView

    @VisibleForTesting
    val mediaControlsView: MediaContainerView?
        get() = mediaContainerController.mediaContainerView

    /** Must be called before use. */
    fun initialize(parent: NotificationStackScrollLayout) {
        check(!initialized) { "NotificationSectionsManager already initialized" }
        initialized = true
        this.parent = parent
        reinflateViews()
        configurationController.addCallback(configurationListener)
    }

    fun createSectionsForBuckets(): Array<NotificationSection> =
        sectionsFeatureManager
            .getNotificationBuckets()
            .map { NotificationSection(parent, it) }
            .toTypedArray()

    /** Reinflates the entire notification header, including all decoration views. */
    fun reinflateViews() {
        silentHeaderController.reinflateView(parent)
        alertingHeaderController.reinflateView(parent)
        peopleHeaderController.reinflateView(parent)
        incomingHeaderController.reinflateView(parent)
        mediaContainerController.reinflateView(parent)
        keyguardMediaController.attachSinglePaneContainer(mediaControlsView)
    }

    override fun beginsSection(view: View, previous: View?): Boolean =
        view === silentHeaderView ||
            view === mediaControlsView ||
            view === peopleHeaderView ||
            view === alertingHeaderView ||
            view === incomingHeaderView ||
            getBucket(view) != getBucket(previous)

    private fun getBucket(view: View?): Int? =
        when {
            view === silentHeaderView -> BUCKET_SILENT
            view === incomingHeaderView -> BUCKET_HEADS_UP
            view === mediaControlsView -> BUCKET_MEDIA_CONTROLS
            view === peopleHeaderView -> BUCKET_PEOPLE
            view === alertingHeaderView -> BUCKET_ALERTING
            view is ExpandableNotificationRow -> view.entry.bucket
            else -> null
        }

    private sealed class SectionBounds {

        data class Many(val first: ExpandableView, val last: ExpandableView) : SectionBounds()

        data class One(val lone: ExpandableView) : SectionBounds()
        object None : SectionBounds()

        fun addNotif(notif: ExpandableView): SectionBounds =
            when (this) {
                is None -> One(notif)
                is One -> Many(lone, notif)
                is Many -> copy(last = notif)
            }

        fun updateSection(section: NotificationSection): Boolean =
            when (this) {
                is None -> section.setFirstAndLastVisibleChildren(null, null)
                is One -> section.setFirstAndLastVisibleChildren(lone, lone)
                is Many -> section.setFirstAndLastVisibleChildren(first, last)
            }

        private fun NotificationSection.setFirstAndLastVisibleChildren(
            first: ExpandableView?,
            last: ExpandableView?
        ): Boolean {
            val firstChanged = setFirstVisibleChild(first)
            val lastChanged = setLastVisibleChild(last)
            return firstChanged || lastChanged
        }
    }

    /**
     * Updates the boundaries (as tracked by their first and last views) of the priority sections.
     *
     * @return `true` If the last view in the top section changed (so we need to animate).
     */
    fun updateFirstAndLastViewsForAllSections(
        sections: Array<NotificationSection>,
        children: List<ExpandableView>
    ): Boolean {
        // Create mapping of bucket to section
        val sectionBounds =
            children
                .asSequence()
                // Group children by bucket
                .groupingBy {
                    getBucket(it)
                        ?: throw IllegalArgumentException("Cannot find section bucket for view")
                }
                // Combine each bucket into a SectionBoundary
                .foldToSparseArray(
                    SectionBounds.None,
                    size = sections.size,
                    operation = SectionBounds::addNotif
                )

        // Build a set of the old first/last Views of the sections
        val oldFirstChildren = sections.mapNotNull { it.firstVisibleChild }.toSet().toMutableSet()
        val oldLastChildren = sections.mapNotNull { it.lastVisibleChild }.toSet().toMutableSet()

        // Update each section with the associated boundary, tracking if there was a change
        val changed =
            sections.fold(false) { changed, section ->
                val bounds = sectionBounds[section.bucket] ?: SectionBounds.None
                val isSectionChanged = bounds.updateSection(section)
                isSectionChanged || changed
            }

        val newFirstChildren = sections.mapNotNull { it.firstVisibleChild }
        val newLastChildren = sections.mapNotNull { it.lastVisibleChild }

        // Update the roundness of Views that weren't already in the first/last position
        newFirstChildren.forEach { firstChild ->
            val wasFirstChild = oldFirstChildren.remove(firstChild)
            if (!wasFirstChild) {
                val notAnimatedChild = !notificationRoundnessManager.isAnimatedChild(firstChild)
                val animated = firstChild.isShown && notAnimatedChild
                firstChild.requestTopRoundness(1f, SECTION, animated)
            }
        }
        newLastChildren.forEach { lastChild ->
            val wasLastChild = oldLastChildren.remove(lastChild)
            if (!wasLastChild) {
                val notAnimatedChild = !notificationRoundnessManager.isAnimatedChild(lastChild)
                val animated = lastChild.isShown && notAnimatedChild
                lastChild.requestBottomRoundness(1f, SECTION, animated)
            }
        }

        // The Views left in the set are no longer in the first/last position
        oldFirstChildren.forEach { noMoreFirstChild ->
            noMoreFirstChild.requestTopRoundness(0f, SECTION)
        }
        oldLastChildren.forEach { noMoreLastChild ->
            noMoreLastChild.requestBottomRoundness(0f, SECTION)
        }

        if (DEBUG) {
            logSections(sections)
        }
        return changed
    }

    private fun logSections(sections: Array<NotificationSection>) {
        for (i in sections.indices) {
            val s = sections[i]
            val fs =
                when (val first = s.firstVisibleChild) {
                    null -> "(null)"
                    is ExpandableNotificationRow -> first.entry.key
                    else -> Integer.toHexString(System.identityHashCode(first))
                }
            val ls =
                when (val last = s.lastVisibleChild) {
                    null -> "(null)"
                    is ExpandableNotificationRow -> last.entry.key
                    else -> Integer.toHexString(System.identityHashCode(last))
                }
            Log.d(TAG, "updateSections: f=$fs s=$i")
            Log.d(TAG, "updateSections: l=$ls s=$i")
        }
    }

    fun setHeaderForegroundColors(@ColorInt onSurface: Int, @ColorInt onSurfaceVariant: Int) {
        peopleHeaderView?.setForegroundColors(onSurface, onSurfaceVariant)
        silentHeaderView?.setForegroundColors(onSurface, onSurfaceVariant)
        alertingHeaderView?.setForegroundColors(onSurface, onSurfaceVariant)
    }

    companion object {
        private const val TAG = "NotifSectionsManager"
        private const val DEBUG = false
        private val SECTION = SourceType.from("Section")
    }
}
